File: longest-increasing-subsequence/lis.py
from typing import List

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        dp = [1] * len(nums)
        maxlen = 0
        for i in range(1, len(nums)):
            for j in range(i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], dp[j]+1)
                    maxlen = max(maxlen, dp[i])
        return maxlen

S = Solution()
print(S.lengthOfLIS([10,9,2,5,3,7,101,18]))

File: longest-increasing-subsequence/lis.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> dp(nums.size(), 1);
        int maxlen = 1;
        for (int i = 1; i < nums.size(); i++) {
            for (int j = 0; j < i; j++) {
                if (nums[j] < nums[i]) {
                    dp[i] = std::max(dp[i], dp[j]+1);
                    maxlen = std::max(maxlen, dp[i]);
                }
            }
        }
        return maxlen;
    }
};
